src/ray/gcs/gcs_server/gcs_actor_manager.h

```diff
@@ -0,0 +1,163 @@
+  /// Get the state of this gcs actor.
+  rpc::ActorTableData_ActorState GetState() const;
+
+  /// Get the id of this actor.
+  ActorID GetActorID() const;
+  /// Get the task specification of this actor.
+  TaskSpecification GetCreationTaskSpecification() const;
+
+  /// Get the immutable ActorTableData of this actor.
+  const rpc::ActorTableData &GetActorTableData() const;
+  /// Get the mutable ActorTableData of this actor.
+  rpc::ActorTableData *GetMutableActorTableData();
+
+  /// Flush the current actor meta info to storage asynchronously.
+  ///
+  /// \param callback Invoked when flushed successfully.
+  void AsyncFlush(std::function<void()> &&callback = nullptr);
+
+ private:
+  /// The actor meta data which contains the task specification as well as the state of
+  /// the gcs actor and so on (see gcs.proto).
+  rpc::ActorTableData actor_table_data_;
+  /// The client to communicate with the redis/storage, it will be replaced with
+  /// StorageClient in the future.
+  std::shared_ptr<RedisGcsClient> redis_gcs_client_;
+};
+
+/// GcsActorManager is responsible for managing the lifecycle of all actors.
+class GcsActorManager {
+ public:
+  /// Create a GcsActorManager
+  ///
+  /// \param io_context The main event loop.
+  /// \param gcs_client Communicate with storage(redis), it will be replaced with
+  /// StorageClient later.
+  /// \param gcs_node_manager The actor manager needs to listen the changes of node
+  /// inside gcs_node_manager.
+  explicit GcsActorManager(boost::asio::io_context &io_context,
+                           std::shared_ptr<gcs::RedisGcsClient> gcs_client,
+                           gcs::GcsNodeManager &gcs_node_manager);
+  ~GcsActorManager() = default;
+
+  /// Register actor asynchronously.
+  ///
+  /// \param request Contains the meta info to create the actor.
+  /// \param callback Will be invoked after the meta info is flushed to the storage or be
+  /// invoked immediately if the meta info is already exist.
+  void RegisterActor(const rpc::CreateActorRequest &request,
+                     std::function<void(std::shared_ptr<GcsActor>)> &&callback);
+
+  /// Reconstruct all actors associated with the specified node id.
+  /// The related actors may be scheduling or have been created.
+  ///
+  /// \param node_id The specified node id.
+  void ReconstructActorsOnNode(const ClientID &node_id);
+
+  /// Reconstruct actor associated with the specified node_id and worker_id.
+  /// The actor may be scheduling or has been created.
+  ///
+  /// \param node_id ID of the node where the worker is located
+  /// \param worker_id  ID of the worker that the actor is creating/created on
+  /// \param no_reconstruct Weather to reconstruct, sometimes users want to kill an actor
+  /// intentionally and don't want it to be reconstructed again.
+  void ReconstructActorOnWorker(const ClientID &node_id, const WorkerID &worker_id,
+                                bool no_reconstruct = false);
```

Comment:
Maybe rename this function or the `no_reconstruct` flag? At first glance, it appears that if `no_reconstruct` is false then this function shouldn't be called.